### PR TITLE
Connector Custom Invoker

### PR DESCRIPTION
**Pull request by LucGenetier**

### CODE REVIEW

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)

File: src/libraries/Microsoft.PowerFx.Interpreter/Environment/SymbolExtensions.cs [hunk 1]
```diff
@@ -41,9 +41,10 @@ public static void SetCulture(this RuntimeConfig symbols, CultureInfo culture)
         /// <param name="symbols">SymbolValues where to set the CultureInfo.</param>
         /// <param name="clock">service to provide current time.</param>
         /// <exception cref="ArgumentNullException">When clock is null.</exception>
-        public static void SetClock(this RuntimeConfig symbols, IClockService clock)
+        public static RuntimeConfig SetClock(this RuntimeConfig symbols, IClockService clock)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Environment/SymbolExtensions.cs:44]: >RuntimeConfig [](http://example.com/codeflow?start=22&length=13)

let's not - 
the other similar APIs in this file don't do this; and this is also an API breaking change.

File: src/libraries/Microsoft.PowerFx.Connectors/Public/BaseRuntimeConnectorContext.cs [hunk 2]
```diff
@@ -2,22 +2,23 @@
...
-using SharpYaml.Serialization.Logging;
+using Microsoft.PowerFx.Connectors.Execution;
 
 namespace Microsoft.PowerFx.Connectors
 {
     public abstract class BaseRuntimeConnectorContext
-    {
-        public abstract HttpMessageInvoker GetInvoker(string @namespace);
+    {   
+        public abstract object GetInvoker(string @namespace);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Public/BaseRuntimeConnectorContext.cs:11]: >object [](http://example.com/codeflow?start=24&length=6)

?

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)<br />❌CP0005 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(Microsoft.PowerFx.Connectors.ConnectorFunction,System.Boolean)

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancellationToken);
+
+        // resquest: support for paging
+        Task<FormulaValue> InvokeAsync(string nextlink, CancellationToken cancellationToken);        
+        
+        // processing/inner call
+        Task<FormulaValue> SendAsync(InvokerParameters invokerElements, CancellationToken cancellationToken);
+    }
+
+    public class InvokerParameters
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs:27]: >InvokerParameters [](http://example.com/codeflow?start=17&length=17)

This is effectively our HttpRequestMessage abstraction. 

It's actually useful to have this.

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        BaseRuntimeConnectorContext Context { get; }
+
+        // request: classic call
+        Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancellationToken);
+
+        // resquest: support for paging
+        Task<FormulaValue> InvokeAsync(string nextlink, CancellationToken cancellationToken);        
+        
+        // processing/inner call
+        Task<FormulaValue> SendAsync(InvokerParameters invokerElements, CancellationToken cancellationToken);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs:24]: >SendAsync [](http://example.com/codeflow?start=27&length=9)

Why do we need both InvokeAsync and SendAsync? Is one layered on the other?

Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs:24]: InvokeAsync is high level and transforms Fx parameters to InvokerParameters. This is where we apply the "policy", aligning the input parameters with what the swagger file mandates.

SendAsync transforms InvokerParameters to a result (FormulaValue) and that's the only abstract method in FunctionInvoker.

It is implemented in HttpFunctionInvoker (where it's the only API being implemented, generating an HttpRequestMessage, getting an HttpResponse and transforming it to a FormulaValue)

This is the key part that allows us to have a custom invoker
You see this in CustomInvokerTests.cs, in CustomFunctionInvoker class, which is making a call to CustomInvoker, itself an equivalent of HttpClient

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs (hunk 3 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs:27]: Yes.

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs (hunk 4 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs:24]: Notice that this approach could help us in our tests, in not using an HttpClient

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    public class InvokerParameters
+    {
+        public QueryType QueryType { get; init; }
+
+        public HttpMethod HttpMethod { get; init; }
+
+        public string Url { get; init; }
+
+        public Dictionary<string, string> Headers { get; init; }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs:35]: >Dictionary [](http://example.com/codeflow?start=15&length=10)

super nit - IReadOnlyDictionary?

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs [hunk 6]
```diff
@@ -1,325 +1,56 @@
...
-        private readonly bool _returnRawResults;
-        private readonly ConnectorLogger _logger;
-
-        public HttpFunctionInvoker(ConnectorFunction function, BaseRuntimeConnectorContext runtimeContext)
-        {
-            _function = function;
-            _httpClient = runtimeContext.GetInvoker(function.Namespace);
-            _returnRawResults = runtimeContext.ReturnRawResults;
-            _logger = runtimeContext.ExecutionLogger;
+    internal class HttpFunctionInvoker : FunctionInvoker
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs:14]: >HttpFunctionInvoker [](http://example.com/codeflow?start=19&length=19)

where do we allocate this?

File: src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs [hunk 7]
```diff
@@ -590,21 +591,22 @@ internal async Task<FormulaValue> InvokeInternalAsync(FormulaValue[] arguments,
             EnsureInitialized();
             runtimeContext.ExecutionLogger?.LogDebug($"Entering in {this.LogFunction(nameof(InvokeInternalAsync))}, with {LogArguments(arguments)}");
             BaseRuntimeConnectorContext context = ConnectorReturnType.Binary ? runtimeContext.WithRawResults() : runtimeContext;
-            ScopedHttpFunctionInvoker invoker = new ScopedHttpFunctionInvoker(DPath.Root.Append(DName.MakeValid(Namespace, out _)), Name, Namespace, new HttpFunctionInvoker(this, context), context.ThrowOnError);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs:593]: >new HttpFunctionInvoker [](http://example.com/codeflow?start=149&length=23)

this is where HttpFunctionInvoker was created before. 
who creates it now?

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs (hunk 6 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs:14]: BaseRuntimeConnectorContext.GetInvoker is where we should have this constructor
Should be virtual and not abstract

File: src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs (hunk 7 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs:593]: Will be in BaseRuntimeConnectorContext.GetInvoker & I'll fix it

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    public interface IConnectorInvoker
+    {
+        // runtime context
+        BaseRuntimeConnectorContext Context { get; }
+
+        // request: classic call
+        Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancellationToken);
+
+        // resquest: support for paging
+        Task<FormulaValue> InvokeAsync(string nextlink, CancellationToken cancellationToken);
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs:21]: To remove from interface

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Connectors.Execution
+{
+    public interface IConnectorInvoker
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs:12]: >IConnectorInvoker [](http://example.com/codeflow?start=21&length=17)

per offline discussion ... 

Consider making this interface internal or removing. 

If everybody actually just derives from FunctionInvoker, we can just use base class. 
And it sounds like users only override SendAsync, they don't override InvokeAsync, Context. So a user doesn't need this interface directly.

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        public string Url { get; init; }
+
+        public IReadOnlyDictionary<string, string> Headers { get; init; }
+
+        public string Body { get; init; }
+
+        public string ContentType { get; init; }        
+    }
+
+    public enum QueryType
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Execution/IConnectorInvoker.cs:24]: >QueryType [](http://example.com/codeflow?start=16&length=9)

Why do we need this at InvokeParameters?
Can these concerns already be abstracted away prior to creating the InvokerParameters?

Can this be removed?

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/IConvertToUTC.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Connectors.Execution
+{
+    internal interface IConvertToUTC
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Execution/IConvertToUTC.cs:9]: >IConvertToUTC [](http://example.com/codeflow?start=23&length=13)

possibly related, see RuntimeConfig.SetTimeZone.

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.#ctor<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_ExecutionLogger<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_TimeZoneInfo<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)<br />❌CP0005 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(Microsoft.PowerFx.Connectors.ConnectorFunction,System.Boolean)

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.#ctor<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_ExecutionLogger<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_TimeZoneInfo<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)<br />❌CP0005 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(Microsoft.PowerFx.Connectors.ConnectorFunction,System.Boolean)

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.#ctor<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_ExecutionLogger<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_TimeZoneInfo<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)<br />❌CP0005 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(Microsoft.PowerFx.Connectors.ConnectorFunction,System.Boolean)

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.#ctor<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_ExecutionLogger<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_TimeZoneInfo<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)<br />❌CP0005 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(Microsoft.PowerFx.Connectors.ConnectorFunction,System.Boolean)

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.#ctor<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_ExecutionLogger<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_TimeZoneInfo<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)<br />❌CP0005 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(Microsoft.PowerFx.Connectors.ConnectorFunction,System.Boolean)

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.#ctor<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_ExecutionLogger<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.get_TimeZoneInfo<br />❌CP0002 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(System.String)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.get_IgnoreUnknownExtensions<br />❌CP0002 M:Microsoft.PowerFx.Connectors.ConnectorSettings.set_IgnoreUnknownExtensions(System.Boolean)<br />❌CP0005 M:Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext.GetInvoker(Microsoft.PowerFx.Connectors.ConnectorFunction,System.Boolean)

**Pull request closed without merging**